Mathlib/RingTheory/Length.lean

```diff
@@ -0,0 +1,123 @@
+  rw [Module.coe_length, Order.coheight_bot_eq_krullDim]
+
+variable {R M}
+
+lemma Module.length_eq_zero_iff : Module.length R M = 0 ↔ Subsingleton M := by
+  rw [← WithBot.coe_inj, Module.coe_length, WithBot.coe_zero,
+    Order.krullDim_eq_zero_iff_of_orderTop, Submodule.subsingleton_iff]
+
+@[nontriviality]
+lemma Module.length_eq_zero [Subsingleton M] : Module.length R M = 0 :=
+  Module.length_eq_zero_iff.mpr ‹_›
+
+lemma Module.length_pos_iff : 0 < Module.length R M ↔ Nontrivial M := by
+  rw [pos_iff_ne_zero, ne_eq, Module.length_eq_zero_iff, not_subsingleton_iff_nontrivial]
+
+lemma Module.length_pos [Nontrivial M] : 0 < Module.length R M :=
+  Module.length_pos_iff.mpr ‹_›
+
+lemma Module.length_compositionSeries (s : CompositionSeries (Submodule R M)) (h₁ : s.head = ⊥)
+    (h₂ : s.last = ⊤) : s.length = Module.length R M := by
+  have H := isFiniteLength_of_exists_compositionSeries ⟨s, h₁, h₂⟩
+  have := (isFiniteLength_iff_isNoetherian_isArtinian.mp H).1
+  have := (isFiniteLength_iff_isNoetherian_isArtinian.mp H).2
+  rw [← WithBot.coe_inj, Module.coe_length]
+  apply le_antisymm
+  · let s' := s.map (β := Submodule R M) (s := (· < ·)) ⟨id, fun h ↦ h.1⟩
+    exact (Order.LTSeries.length_le_krullDim s')
+  · rw [Order.krullDim, iSup_le_iff]
+    intro t
+    refine WithBot.coe_le_coe.mpr ?_
+    obtain ⟨t', i, hi, ht₁, ht₂⟩ := t.exists_relSeries_covBy_and_head_eq_bot_and_last_eq_bot
+    have := (s.jordan_holder t' (h₁.trans ht₁.symm) (h₂.trans ht₂.symm)).choose
+    have h : t.length ≤ t'.length := by simpa using Fintype.card_le_of_embedding i
+    have h' : t'.length = s.length := by simpa using (Fintype.card_congr this.symm)
+    simpa using h.trans h'.le
+
+lemma Module.length_ne_top_iff : Module.length R M ≠ ⊤ ↔ IsFiniteLength R M := by
+  constructor
+  · nontriviality M
+    cases finiteDimensionalOrder_or_infiniteDimensionalOrder (Submodule R M)
+    · intro _
+      rw [isFiniteLength_iff_isNoetherian_isArtinian, isNoetherian_iff, isArtinian_iff]
+      exact ⟨Rel.wellFounded_swap_of_finiteDimensional _, Rel.wellFounded_of_finiteDimensional _⟩
+    · rw [ne_eq, ← WithBot.coe_inj, Module.coe_length, WithBot.coe_top, Order.krullDim_eq_top]
+      simp only [not_true_eq_false, IsEmpty.forall_iff]
+  · intro H
+    obtain ⟨s, hs₁, hs₂⟩ := isFiniteLength_iff_exists_compositionSeries.mp H
+    rw [← length_compositionSeries s hs₁ hs₂]
+    simp
+
+lemma Module.length_ne_top [IsArtinian R M] [IsNoetherian R M] : Module.length R M ≠ ⊤ := by
+  rw [Module.length_ne_top_iff, isFiniteLength_iff_isNoetherian_isArtinian]
+  exact ⟨‹_›, ‹_›⟩
+
+variable {N P : Type*} [AddCommGroup N] [AddCommGroup P] [Module R N] [Module R P]
+variable (f : N →ₗ[R] M) (g : M →ₗ[R] P) (hf : Function.Injective f) (hg : Function.Surjective g)
+variable (H : Function.Exact f g)
+
+include hf hg H in
+/-- Length is additive in exact sequences. -/
+lemma Module.length_eq_add_of_exact :
+    Module.length R M = Module.length R N + Module.length R P := by
```

Comment:
If you prove how module length behaves under injective / surjective linear maps and isomorphisms, you can get this from the case where P=M/N, which is probably easier ...
(we have [Submodule.mapIic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Submodule/Range.html#Submodule.mapIic) (and [Order.height_eq_krullDim_Iic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/KrullDimension.html#Order.height_eq_krullDim_Iic)) but apparently not the quotient module / `Ici` version)

Reply:
I don't see immediately how this would be easier. You still have to show that the submodule lattice is "catenary" (i.e. krulldim = max length of series through a fixed element)

Reply:
Indeed, it appears that we still need to use `map` and `comap` to construct the two series that are smashed together. However it would be nice to have the inequalities between lengths induced by injection/surjection.

Reply:
Indeed. I planned to do so but forgot. Here's the follow-up PR: #23679